File: server/bs_rpc/msg/msg.go
package msg

import (
    "context"
)

const (
    METHOD_CALL_FOR_RESPONSE  = "call_for_response"
    METHOD_CALL_FOR_RESPONSES = "call_for_responses"
    METHOD_RESPONSE_CLOSE     = "response close"
    METHOD_RESPONSE           = "response"
)

type SendClient interface {
    Send(body Body) error
}

type Responder func(body Body) error

type Type string
type Method string

type Body interface {
    CallForResponse(ctx context.Context) (Body, error)
    CallForResponses(ctx context.Context, r Responder) error
    GetType() Type
}

type Msg[T Body] struct {
    Type   Type   `json:"type"`
    Method Method `json:"method"`
    Call   uint32 `json:"call"`
    Body   T      `json:"body"`
}

func NewMsg[T Body](b T, call uint32, method Method) Msg[T] {
    return Msg[T]{b.GetType(), method, call, b}
}

File: bs_rpc/msg.py
from enum import Enum
from typing import TypedDict


class Method(Enum):
    CALL_FOR_RESPONSE = "call_for_response"
    CALL_FOR_RESPONSES = "call_for_responses"
    RESPONSE_CLOSE = "response close"
    RESPONSE = "response"


class Body:
    def __init__(self, msg_type: str, data: dict):
        self.type = msg_type
        self.data = data

    def call_for_response(self, context):
        """returns Body"""
        raise Exception(f"call_for_response not implemented for {self.type}")

    def call_for_responses(self, context, response_writer):
        raise Exception(f"call_for_responses not implemented for {self.type}")

    def get_type(self):
        return self.type


class Msg(TypedDict):
    type: str
    method: str
    call: int
    body: dict


class EmptyMsg(Body):
    def __init__(self, _):
        super().__init__("internal.empty", {})
